### PR TITLE
add gradient for broadcast_to

### LINKED ISSUE #21901: Add gradient for tf.broadcast_to
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**: binary (pip tensorflow-gpu)
- **TensorFlow version (use command below)**: v1.10.0-0-g656e7a2b34 1.10.0
- **Python version**: 3.6.6
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: CUDA 9.0 / CUDNN 7.1.3
- **GPU model and memory**:
- **Exact command to reproduce**:
```
python -c "import tensorflow as tf; a = tf.zeros([1]); b = tf.broadcast_to(a, [2]); tf.gradients(b, a)"
```
Result:
LookupError: gradient registry has no entry for: BroadcastTo

### Describe the problem
Problem: tf.broadcast_to (BroadcastTo) does not have a gradient defined (on CPU or GPU)
Feature request: Add the gradient for the op.
Relevant links:
https://github.com/tensorflow/tensorflow/pull/19753
https://github.com/tensorflow/tensorflow/pull/15243

### Source code / logs
```
Traceback (most recent call last):
  File "/h/username/.local/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 733, in _GradientsHelper
    grad_fn = ops.get_gradient_function(op)
  File "/h/username/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 2335, in get_gradient_function
    return _gradient_registry.lookup(op_type)
  File "/h/username/.local/lib/python3.6/site-packages/tensorflow/python/framework/registry.py", line 93, in lookup
    "%s registry has no entry for: %s" % (self._name, name))
LookupError: gradient registry has no entry for: BroadcastTo

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/h/username/.local/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 596, in gradients
    gate_gradients, aggregation_method, stop_gradients)
  File "/h/username/.local/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 737, in _GradientsHelper
    (op.name, op.type))
LookupError: No gradient defined for operation 'BroadcastTo' (op type: BroadcastTo)
```

### PR DESCRIPTION
**Pull request by facaiy**
Fix #21901

### CODE REVIEW

**Comment by facaiy** [comment]
@alextp Could you take a  look?

File: tensorflow/python/kernel_tests/broadcast_to_ops_test.py [hunk 1]
```diff
@@ -81,5 +82,24 @@ def testBroadcastToShapeTypeAndInference(self):
         # check shape inference when shape input is constant
         self.assertAllEqual(shape, v_np.shape)
 
+  def testGradient(self):
```
Comment (alextp) [inline tensorflow/python/kernel_tests/broadcast_to_ops_test.py]: Can you add a few more tests?

Specifically, the gradient for broadcasting from a scalar to a vector (increasing rank). Also the case of broadcasting when the rank does not increase (say from shape [2, 1, 3] to shape [2, 5, 3]). Also one case which covers both (broadcasting from shape [2, 1] to shape [5, 2, 3]).

**Review by alextp — CHANGES_REQUESTED** [review]
Thanks for the pull request!

While I think an implementation based on reduce will be faster this one is better than having no gradient for broadcast at all, so I'm happy to merge this now and optimize later.

I just want a few more tests.

File: tensorflow/python/kernel_tests/broadcast_to_ops_test.py (hunk 1 above)
Comment (facaiy) [reply tensorflow/python/kernel_tests/broadcast_to_ops_test.py]: Sure. Is it good to register gradient for C++ side as well?

Comment (facaiy) [reply tensorflow/python/kernel_tests/broadcast_to_ops_test.py]: By the way, do we need to create a new file for gradient, say broadcast_to_ops_grad_test.py ?

Comment (alextp) [reply tensorflow/python/kernel_tests/broadcast_to_ops_test.py]: No need to make a new file or register C++ gradients.

**Comment by facaiy** [comment]
Hi, more tests are added. Could you take a look again? Thanks.

**Review by alextp — DISMISSED** [review]

**Comment by alextp** [comment]
It seems like one of these tests is not safe on the GPU. I think there is a bug with broadcast_to on GPU from scalars; can you make that test cpu-only by adding a `with ops.device("cpu:0"):` block around it?

**Comment by facaiy** [comment]
I have made this scalar test cpu-only, but I didn't test it on my machine. Could you restart all tests? Thanks.

**Review by alextp — APPROVED** [review]

**Comment by facaiy** [comment]
@alextp Thanks for your review. I think those failed tests are unrelated.

**Comment by alextp** [comment]
Indeed, they look like flakes. Rerunning.

**Comment by facaiy** [comment]
Thank you, @alextp.

**Pull request merged**